### ISSUE TITLE
Add to query service dataset rather than replace

### ISSUE THREAD

**Issue by kjetilk** [w3c/sparql-12#95]
## Why?

I've been brainstorming [an idea](https://github.com/solid/solid-spec/issues/186), where I'd like the query engine to form the default in various ways, but where people can add graphs to the elaborately designed dataset as needed. However, the current spec says:

> A SPARQL query may specify the dataset to be used for matching by using the FROM clause and the FROM NAMED clause to describe the RDF dataset. If a query provides such a dataset description, then it is used in place of any dataset that the query service would use if no dataset description is provided in a query.

So, it prevents that, as any `FROM` clause overrides that default graph.

## Previous work

I'm not aware of anything.

## Proposed solution

How about: 

If a query provides such a dataset description, then it is used in place of any dataset that the query service would use if no dataset description is provided in a query, unless an `ADD` keyword is provided. For example:

```sparql
ADD FROM <http://example.org/foobar>
```
will add to the service-supplied default graph, not replace it.

## Considerations for backward compatibility

It should be fully backwards compatible.

**Comment by namedgraph** [comment]
Isn't this what [`LOAD`](https://www.w3.org/TR/sparql11-update/#load) does?

**Comment by kjetilk** [comment]
> Isn't this what [`LOAD`](https://www.w3.org/TR/sparql11-update/#load) does?

Not really. `LOAD` is about dereferencing a document and give it a graph name. This issue is about how you build the RDF Dataset from already named "local" graphs.

**Comment by namedgraph** [comment]
Then it looks to me that this is about the [union graph](#59)

**Comment by kasei** [comment]
> Then it looks to me that this is about the union graph #59

It seems related, but I'm not sure it could do exactly what @kjetilk wants. My understanding of the issue here is for endpoints who have a specific union graph set up as the default graph, but it specifically does *not* include *all* graphs. To that union, you'd like to add some fixed set of other named graphs to construct the default graph for your request.

For example, assume an endpoint has 5 named graphs: A, B, C, D, E, and that it is configured to have the default graph be the union of A and B. How do you make a request such that the default graph is A, B, and C? Two possibilities occur to me:

1. Having enough information in the service description so that clients could reconstruct the default entirely and then add whatever they want to it
2. @kjetilk's suggestion that there be a way to add on to the default

1 seems doable right now, since it only involves metadata, and a smarter client. But 2 certainly seems more convenient in situations where the default might be a very large union, and it avoids the extra round-trip of requesting the service description, parsing it, and then adding lots of extra `FROM` clauses to your request.

@kjetilk – did I understand that right?

**Comment by kjetilk** [comment]
> @kjetilk – did I understand that right?

Yes, that was an accurate description!

**Comment by lisp** [comment]
why not be able to designate that "graph the constitution of which is not known"?

**Comment by kjetilk** [comment]
I didn't understand your question, @lisp

**Comment by lisp** [comment]
why require a description based on which the client can construct this non-trivial "default" graph by reflecting that description back to the service when the same could be achieve by just giving it a name?

**Comment by kjetilk** [comment]
I still don't understand your question, @lisp ... Are you thinking about @kasei 's suggestion 1. above, where the client is given a list of graphs in a service description and is thus able to explicitly name all graphs for inclusion in the default graph? 

If so, then, no, that list may be really large and would complicate formulate the query writing, the point of the exercise is to make writing the query simple for a human.

**Comment by lisp** [comment]
you did understand my question. 

>  the point of the exercise is to make writing the query simple for a human.

that is my concern.

**Comment by kjetilk** [comment]
Right, so basically, defining the default graph could be the concern of the server, and only in exceptional cases, the concern of the client. Then, it is better to have a simple mechanism to add rather than a mandate to have to deal with the entire problem on the client side.
